File: Week1 C Problems/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    long number;
    do
    {
        number = get_long("Enter your card number: ");
    }
    while (number < 0);

    int sum = 0;
    int count = 0;
    long n = number;

    while (n > 0)
    {
        int digit = n % 10;

        if (count % 2 == 0)
        {
            sum += digit;
        }
        else
        {
            int doubled = digit * 2;
            if (doubled > 9)
            {
                sum += (doubled % 10) + (doubled / 10);
            }
            else
            {
                sum += doubled;
            }
        }

        n /= 10;
        count++;
    }

    long start = number;
    while (start >= 100)
    {
        start /= 10;
    }

    if (sum % 10 == 0)
    {
        if ((start == 34 || start == 37) && count == 15)
        {
            printf("AMEX\n");
        }
        else if ((start >= 51 && start <= 55) && count == 16)
        {
            printf("MASTERCARD\n");
        }
        else if ((start / 10 == 4) && (count == 13 || count == 16))
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }

    return 0;
}

File: Week6 Python Problems/credit.py
# credit.py
from cs50 import get_string

def luhn_checksum(card_number):
    total = 0
    reverse_digits = card_number[::-1]

    for i, digit in enumerate(reverse_digits):
        n = int(digit)
        if i % 2 == 1:  # every second digit from the right (0-indexed from reversed)
            n *= 2
            if n > 9:
                n -= 9
        total += n
    return total % 10 == 0

def main():
    card_number = get_string("Number: ")

    # Validate using Luhn’s Algorithm
    if not luhn_checksum(card_number):
        print("INVALID")
        return

    length = len(card_number)
    first_two = int(card_number[:2])
    first_one = int(card_number[0])

    # Check card type
    if length == 15 and first_two in [34, 37]:
        print("AMEX")
    elif length == 16 and 51 <= first_two <= 55:
        print("MASTERCARD")
    elif length in [13, 16] and first_one == 4:
        print("VISA")
    else:
        print("INVALID")

if __name__ == "__main__":
    main()
